core/src/orderbook/streamed/orderbook.rs

```diff
@@ -111,15 +129,16 @@ impl TryFrom<File> for Orderbook {
     type Error = anyhow::Error;
 
     fn try_from(mut file: File) -> Result<Self> {
-        let mut contents = Vec::new();
-        let bytes_read = file
-            .read_to_end(&mut contents)
+        let orderbook = Orderbook::read(&mut file)
```

Comment:
Can we not impl `TryFrom<Read>` here?

Reply:
No because Rust doesn't support trait specializations:
```
struct Foo;
impl<T: std::io::Read> std::convert::TryFrom<T> for Foo {
    type Error = ();
    fn try_from(path: T) -> Result<Self, Self::Error> {
        Ok(Foo)
    }
}
````

Gives the error:
```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<_>` for type `Foo`:
 --> src/lib.rs:2:1
  |
2 | impl<T: std::io::Read> std::convert::TryFrom<T> for Foo {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `core`:
          - impl<T, U> std::convert::TryFrom<U> for T
            where U: std::convert::Into<T>;
```

Specifically, this is because there is a potential conflicting implementation for all U that implement both `Read` and `Into<Foo>`.

Also, from the `TryFrom` docs:
> Simple and safe type conversions that may fail in a controlled way under some circumstances.

So to me, the `TryFrom` is more for modeling conversion behaviour and not decoding, explaining why there is a separate `FromStr` trait for string parsing and not just `TryFrom<&str>` (lifetime variance differences aside).